docs/performance/remote-filesystem/amazon-s3.md

```diff
@@ -0,0 +1,396 @@
+addition of Flysystem and caching libraries, as well as custom service provider
+code to your project.
+
+1. Set-up, configure, and confirm operational your Amazon S3 bucket
+1. Update `composer.json` in your project root
+1. Creating service providers
+1. Bootstrap your providers
+
+Prerequisites
+-------------
+
+<p class="warning"><strong>Warning:</strong> This documentation, and examples,
+assume you have a fully configured and working Amazon S3 bucket.
+</p>
+
+
+
+### Required Libraries
+----------------------
+
+You need to add the following libraries to your project:
+
+  * `league/flysystem-aws-s3-v3:^1.0`
+  * `league/flysystem-cached-adapter:^1.0`
+  * `league/flysystem-memory:^1.0`
+
+This can simply be achieved by running:
+
+```bash
+$ composer require league/flysystem-aws-s3-v3:^1.0 league/flysystem-cached-adapter:^1.0 league/flysystem-memory:^1.0
+```
+
+
+### Caching Libraries
+----------------------
+
+While other cache targets are optional, right now this example will be base on Redis.
+
+To use Redis, you will need a configured and running Redis instance, and need
+to install `predis/predis:^1.1`, which can be achieved by running:
+
+
+```bash
+$ composer require predis/predis:^1.1
+```
+
+
+Autoloading
+------------
+
+Autoloading of your service providers can be set-up in the same way as
+[Extension Bundles][bundles], by adding your base namespace and source
+directory ot the `psr-4` section of your project root `composer.json` file,
+e.g.
+
+```json
+    "autoload": {
+        "psr-4": {
+            "MySiteApp\\": "src"
+        }
+    }
+```
+
+
+<p class="note"><strong>Note:</strong> If you change anything in the
+<code>psr-4</code> section of your <code>composer.json</code> file, remember to
+re-run <code>composer dump-autoload</code> to update the cached autoloading
+files.
+</p>
+
+
+Service Providers
+-----------------
+
+### Amazon AWS
+
+The provider for Amazon AWS services needs to define:
+  * `$app['aws']` - Returns an instance of `\Aws\Sdk`, used to builds AWS
+    clients based on configuration settings
+  * `app['aws.s3']` - Return an `\Aws\S3\S3Client` used to interact with
+    Amazon Simple Storage Service (Amazon S3)
+
+For detailed information on constructing these classes, and the options
+provided see the [Amazon S3 documentation][s3-docs], and specifically the
+[S3 SDK documentation][s3-sdk] and [S3 ApiGen documentation][s3-api].
+
+
+#### Amazon Credentials
+
+In this example we use `CredentialProvider::defaultProvider()` to get your
+Amazon credentials.
+
+This credential provider first checks in order:
+  * Amazon specific environment variables
+  * "default" profile in ~/.aws/credentials
+  * "profile default" profile in ~/.aws/config (which is the default profile of
+    AWS CLI)
+  * Tries to make a GET Request to fetch credentials if the Ecs environment
+    variable is presented
+  * Checks for EC2 instance profile credentials
```

Comment:
Change to "Fetches credentials for the EC2 instance's IAM Role".

Would also note that "If you are using EC2 instances then this is the recommended way to provide credentials".